src/mlpack/methods/ann/augmented/tree_memory_impl.hpp

```diff
@@ -0,0 +1,119 @@
+/**
+ * @file tree_memory_impl.hpp
+ * @author Konstantin Sidorov
+ *
+ * Implementation of CopyTask class
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_AUGMENTED_TREE_MEMORY_IMPL_HPP
+#define MLPACK_METHODS_AUGMENTED_TREE_MEMORY_IMPL_HPP
+
+#include <cassert>
+
+#include "tree_memory.hpp"
+
+namespace mlpack {
+namespace ann /* Artificial Neural Network */ {
+namespace augmented /* Augmented neural network */ {
+template<typename T, typename J, typename W>
+TreeMemory<T, J, W>::TreeMemory(size_t size, J joiner, W writer) {
+  memorySize = size;
+  // Rounding size to the next highest power of 2.
+  assert(0 < size && size < static_cast<size_t>(1 << 31));
+  actualMemorySize = 1 << static_cast<int>(ceil(log2((double) size)));
+  assert(memorySize <= actualMemorySize);
+  // Allocating enough memory to store all leaf values AND inner node values.
+  memory.resize(2 * actualMemorySize - 1);
+  joinFunction = joiner;
+  writeFunction = writer;
+}
+
+template<typename T, typename J, typename W>
+inline T TreeMemory<T, J, W>::Get(size_t index) {
+  assert(0 <= index && index < memorySize);
+  return memory[actualMemorySize - 1 + index];
+}
+
+template<typename T, typename J, typename W>
+inline T& TreeMemory<T, J, W>::Get(size_t index) {
+  assert(0 <= index && index < memory.size());
+  return memory[index];
+}
+
+template<typename T, typename J, typename W>
+inline T TreeMemory<T, J, W>::GetCell(size_t index) {
+  assert(0 <= index && index < memory.size());
+  return memory[index];
+}
```

Comment:
Well, they're not quite the same - the first one extracts *leafs* and the seconds one extracts any node via "pointer". I do agree that current name for the first function is a misnomer - changed that to `GetLeaf`.